server/src/main/java/com/google/coffeehouse/servlets/UpdateClubServlet.java

```diff
@@ -0,0 +1,197 @@
+   *     with an object that represents the Book the Club is reading. The update mask must be
+   *     associated with the key {@code "updateMask"} and have the value of a comma separated list
+   *     of the fields to be updated from the "club" key. To specify an update inside of the book
+   *     object inside of the Club, prepend the field name in the update mask with "currentBook.".
+   *     If no update mask exists, all fields from the "club" key will be used to update the Club
+   *     (and nested Book). If the request does not have either of these keys, or is syntactically
+   *     incorrect, the response object will send a "400 Bad Request error". If the user attempting
+   *     to update the Club is not the Club owner, the response object will send a "403 Forbidden
+   *     error"
+   * @param response the response from this method, will contain the updated Club in JSON format.
+   *     If the request object does not have a valid JSON body (as described in the request
+   *     parameter) this object will send a "400 Bad Request error". If the user attempting
+   *     to update the Club is not the Club owner, this object will send a "403 Forbidden error"
+   * @throws IOException if an input or output error is detected when the servlet handles the request
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Club club;
+    try {
+      JsonObject requestJson = gson.fromJson(request.getReader(), JsonObject.class);
+      JsonObject updatedClubJson = requestJson.getAsJsonObject(CLUB_FIELD_NAME);
+      JsonObject updatedBookJson = updatedClubJson.getAsJsonObject(Club.CURRENT_BOOK_FIELD_NAME);
+      JsonElement updateMask = requestJson.get(UPDATE_MASK_FIELD_NAME);
+      
+      // Get attempted updates for the Club and the Book inside the Club.
+      List<String> bookMask = new ArrayList<>();
+      List<String> clubMask = new ArrayList<>();
+      if (updateMask == null) {
+        bookMask = updateableBookFields;
+        clubMask = updateableClubFields;
+      } else {
+        // Cannot use stream here because Java will complain about local variables not being final.
+        String[] elements = updateMask.getAsString().split(",");
+        for (String element : elements) {
+          boolean isBookUpdate = element.contains(Club.CURRENT_BOOK_FIELD_NAME + ".");
+          if (isBookUpdate) {
+            bookMask.add(element.split("\\.")[1]);
+          } else {
+            clubMask.add(element);
+          }
+        }
+      }
+
+      JsonElement clubIdElement = updatedClubJson.get(Club.CLUB_ID_FIELD_NAME);
+      if (clubIdElement == null) {
+        throw new IllegalArgumentException(String.format(NO_FIELD_ERROR, Club.CLUB_ID_FIELD_NAME));
+      }
+      club = storageHandler.fetchClubFromId(clubIdElement.getAsString());
+
+      JsonElement userIdElement = requestJson.get(Person.USER_ID_FIELD_NAME);
```

Comment:
Hmm, are we just trusting the frontend to tell us which user they are? If so, is there any way an attacker could figure out another User's ID? This is particularly relevant if the Id we're using is not unique to this app (my guess is it is, but I'm not sure)

Another strategy would be to have the client send up its full cryptographically-signed `userid` payload, that way the server can verify that the payload actually originated from a Google server and was not altered.

Reply:
You're so right! I've refactored the code to take the OpenID generated ID token as an input instead of just the userId. This allows us to verify that the user trying to update the club is actually who they claim to be.